packages/dds/tree/src/test/shared-tree/fuzzEditGenerator.ts

```diff
@@ -0,0 +1,331 @@
+        const trees = state.testTreeProvider.trees;
+        const tree = trees[state.treeIndex];
+
+        // generate edit for that specific tree
+        const {
+            parent: path,
+            field: nodeField,
+            index: nodeIndex,
+        } = getRandomNodePosition(tree, state.random);
+        const insert: FuzzInsert = {
+            fuzzType: "insert",
+            parent: path,
+            field: nodeField,
+            index: nodeIndex,
+            value: state.random.integer(Number.MIN_SAFE_INTEGER, Number.MIN_SAFE_INTEGER),
+        };
+        return insert;
+    }
+
+    async function deleteGenerator(state: EditState): Promise<FuzzDelete> {
+        const trees = state.testTreeProvider.trees;
+        const tree = trees[state.treeIndex];
+        // generate edit for that specific tree
+        const path = containsAtLeastOneNode(tree)
+            ? getExistingRandomNodePosition(tree, state.random)
+            : undefined;
+        return {
+            fuzzType: "delete",
+            path,
+        };
+    }
+
+    async function setPayloadGenerator(state: EditState): Promise<FuzzSetPayload> {
+        const trees = state.testTreeProvider.trees;
+        const tree = trees[state.treeIndex];
+        // generate edit for that specific tree
+        const path = containsAtLeastOneNode(tree)
+            ? getExistingRandomNodePosition(tree, state.random)
+            : undefined;
+        return {
+            fuzzType: "setPayload",
+            path,
+            value: state.random.integer(Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER),
+        };
+    }
+
+    /**
+     * currently the acceptance conditions are based on some arbitrary number of edits.
+     * This should be changed to the max number of nodes allowed in tree.
+     * TODO: create helper function to get number of nodes in tree and use that to set acceptance condition
+     * */
+    const baseEditGenerator = createWeightedAsyncGenerator<FuzzChange, EditState>([
+        [insertGenerator, 5],
+        [deleteGenerator, 1],
+        [setPayloadGenerator, 1],
+    ]);
+
+    return async (state: FuzzTestState): Promise<Operation | typeof done> => {
+        const trees = state.testTreeProvider.trees;
+        // does not include last tree, as we want a passive client
+        const treeIndex = state.random.integer(0, trees.length - 2);
+        const contents = await baseEditGenerator({
+            ...state,
+            treeIndex,
+        });
+        state.numberOfEdits += 1;
+        if (contents === done) {
+            return done;
+        }
+        return { type: "edit", contents, index: treeIndex };
+    };
+};
+
+export function makeOpGenerator(): AsyncGenerator<Operation, FuzzTestState> {
+    const opWeights: AsyncWeights<Operation, FuzzTestState> = [
+        [makeEditGenerator(), 3],
+        [{ type: "synchronize" }, 1],
+    ];
+    return createWeightedAsyncGenerator(opWeights);
+}
+
+export interface NodeLocation {
```

Comment:
I think this structure is now equivalent to `UpPath` so you should be able to use that instead of introducing a new concept.